metadata/src/utils/retain.rs

```diff
@@ -8,45 +8,188 @@ use crate::{
     ExtrinsicMetadata, Metadata, OuterEnumsMetadata, PalletMetadataInner, RuntimeApiMetadataInner,
     StorageEntryType,
 };
-use alloc::collections::BTreeMap;
-use hashbrown::HashSet;
-use scale_info::TypeDef;
-
-/// Collect all type IDs needed to represent the provided pallet.
-fn collect_pallet_types(pallet: &PalletMetadataInner, type_ids: &mut HashSet<u32>) {
-    if let Some(storage) = &pallet.storage {
-        for entry in storage.entries() {
-            match entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    type_ids.insert(ty);
+use alloc::collections::{BTreeMap, BTreeSet, VecDeque};
+use scale_info::{
+    PortableType, TypeDef, TypeDefArray, TypeDefBitSequence, TypeDefCompact, TypeDefComposite,
+    TypeDefSequence, TypeDefTuple, TypeDefVariant,
+};
+
+#[derive(Clone)]
+struct TypeSet {
+    seen_ids: BTreeSet<u32>,
+}
+
+impl TypeSet {
+    fn new() -> Self {
+        Self {
+            seen_ids: BTreeSet::new(),
+        }
+    }
+
+    fn insert(&mut self, id: u32) -> bool {
+        self.seen_ids.insert(id)
+    }
+
+    fn contains(&mut self, id: u32) -> bool {
+        self.seen_ids.contains(&id)
+    }
+
+    /// This function will deeply traverse the inital type and it's dependencies to collect the relevant type_ids
+    fn collect_types(&mut self, metadata: &Metadata, t: &PortableType) {
```

Comment:
One final thing abotu this funciton: I would use a `u32` type ID  instead of `&PortableType` as the input personally; it's the thing we tend to use across the code when referencing types, and I think everything ends up going via `insert_and_collect` anyway which takesa. u32 and resolves it so that it can call this fn

Reply:
done